### PR TITLE
Update media plugin simulation to be consistent with plugin specification

### PR DESCRIPTION
**Pull request by albertinad**
Hi! This PR is a fix for the issue #40:
- Improving error handling in order to follow the specification and send a `MediaError` object.
- Keep track of the current state, to vaildate if it is possible to perform actions such as play, pause and stop.
- Improve the creation and validation of the Audio instance.
- Minor code refactor

Thanks!

### CODE REVIEW

**Comment by albertinad** [comment]
@dlebu @guillaumejenkins @busykai I have changed how the simulation of Media plugin is done. In the old implementation, it was a copy of the browser proxy implementation. After doing some tests, I find out that it doesn't behaves the same in a browser, in a real device and in the simulator. So I find some issues with that, and start fixing them in previous commits. But since we override the `window.Media` object with our simulation, and I have added some others private APIs, happens that we end up providing a different media instance for developers, with private APIs that are not available in a real device. Also we just copy most of the javascript side implementation of the real plugin for simulation, and we need to  keep in sync with the original.
In the latests commit I tried to propose a different approach:
- Do not override the `window.Media` object with something similar than the real JS interface. Just use the exactly JS that the plugin provides and write as a simulation, the native calls to exec.
- The browser implementation is not following the standard, so I based my simulation code with android native implementation
With this changes, the behaviour of the media plugin on the simulator is like in a real device.
For testing it, I use the `mobilespec app` and also a code snippet that was the driver of the changes, since with the same code base, I have really different behaviour in the simulator and in a real device, like android:
```
// add some button or element to interact with

 
function playSound() {
    var onSuccess = function() {
        media.stop();
        media.release();

        console.log("onSuccess");
    };

    var onError = function(error) {
        media.stop();
        media.release();

        console.log("onError", error);
    };

    function onStatus(status) {
        var msg;
        switch(status) {
            case Media.MEDIA_NONE: // 0
                msg = "MEDIA_NONE";
                break;
            case Media.MEDIA_STARTING: // 1
                msg = "MEDIA_STARTING";
                break;
            case Media.MEDIA_RUNNING: // 2
                msg = "MEDIA_RUNNING";
                break;
            case Media.MEDIA_PAUSED: // 3
                msg = "MEDIA_PAUSED";
                break;
            case Media.MEDIA_STOPPED: // 3
                msg = "MEDIA_STOPPED";
                break;
            default:
                msg = "unknown status";
        }
        console.log("onStatus: " + msg, status) ;
    }

    // use some short audio, a few seconds (~2 seconds)
    var src =  'audio/some-short-sound.mp3',
        media = new Media(src, onSuccess, onError, onStatus);

    media.play();
}

// at some point
document.getElementById('play-sound').addEventListener('click', playSound);
```

Now with these changes, I have the same behaviour in a real device and in the simulator.

File: src/plugins/cordova-plugin-media/app-host-handlers.js [hunk 1]
```diff
@@ -2,281 +2,106 @@
...
+            'stopRecordingAudio': function (success, fail, args) {
+                var player = Player.getById(args[0]);
+                if (player) {
+                    player.stopRecord(true);
+                }
+            },
+            'pauseRecordingAudio': function (success, fail, args) {
+                var player = Player.getById(args[0]);
+                if (player) {
+                    player.stopRecord(false);
```
Comment (dlebu) [inline src/plugins/cordova-plugin-media/app-host-handlers.js]: Should the `pause` parameter be `true` here and `false` in the `stopRecordingAudio` call?

**Comment by dlebu** [comment]
This looks good to me, @albertinad. Thanks for doing your due diligence to improve the implementation!

File: src/plugins/cordova-plugin-media/app-host-handlers.js (hunk 1 above)
Comment (albertinad) [inline src/plugins/cordova-plugin-media/app-host-handlers.js]: According to the implementation [here](https://github.com/apache/cordova-plugin-media/blob/master/src/android/AudioHandler.java#L120), this should be `true` for `stopRecordingAudio` and `false` for `pauseRecordingAudio`. This is just to follow the other implementation, but when we support recording, we can change it :)

**Comment by albertinad** [comment]
@dlebu No problem! :) thanks!

**Comment by albertinad** [comment]
@dlebu Is it ok if I merge this?

**Comment by dlebu** [comment]
@albertinad sure! :shipit:

**Comment by albertinad** [comment]
Merged! thanks!

**Pull request merged**